### PR TITLE
fix(core): memory leak if view container host view is destroyed while view ref is not

### LINKED ISSUE #38648: bug(core): viewRef._viewContainerRef can leak memory (ViewRef, ViewContainerRef)
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core

### Is this a regression?

This is a problem in Ivy and View Engine.

### Description

The `attachToViewContainerRef()` method of the `ViewRef` class is called whenever the `ViewRef` is attached to a `ViewContainerRef`, and this `ViewContainerRef` instance is assigned to `viewRef._viewContainerRef`.

https://github.com/angular/angular/blob/930eeaf177a4c277f437f42314605ff8dc56fc82/packages/core/src/render3/view_ref.ts#L284-L289

The **only** place where the `_viewContainerRef` property is cleared is when `viewRef.destroy()` is manually called.

https://github.com/angular/angular/blob/930eeaf177a4c277f437f42314605ff8dc56fc82/packages/core/src/render3/view_ref.ts#L75-L88

When `viewContainerRef.detach()` is used to detach a `ViewRef`, the `_viewContainerRef` property is still referencing the container from which it was detached. If this container should be destroyed (e.g. a `ViewContainerRef` of a component being destroyed), then the `_viewContainerRef` property will continue to hold a reference to the destroyed container, causing a memory leak (until the next time this `ViewRef` is attached to a different container).

----

Upon initial glance of the source code, I was not able to find a quick solution to this issue. I don't know if there is a way to access a `ViewRef` instance inside the node_manipulation `detachView()` function.

https://github.com/angular/angular/blob/930eeaf177a4c277f437f42314605ff8dc56fc82/packages/core/src/render3/node_manipulation.ts#L319

Essentially, there needs to be some type of `detachFromViewContainerRef()` method to complement the `attachToViewContainerRef()` method.

## 🔬 Minimal Reproduction

[StackBlitz example](https://stackblitz.com/edit/angular-ymlxv8?file=src%2Fapp%2Fapp.component.ts)
What's happening here:
1. Manually create an `EmbeddedViewRef`.
2. Manually create a dynamic component.
3. Attach the `EmbeddedViewRef` to the dynamic component's `ViewContainerRef`.
4. Detach the `EmbeddedViewRef`.
5. Destroy the dynamic component.

----

(Using Chrome DevTools)
- Load the StackBlitz example.
- Take heap snapshot before doing anything.
	- Search objects for `ViewContainerRef` and you'll find **two** instances.
- Click "create and attach" button.
- Take heap snapshot.
	- Search objects for `ViewContainerRef` and you'll find **three** instances.
- Click "detach and destroy" button.
- Take heap snapshot.
	- Search objects for `ViewContainerRef` and you'll find **three** instances.
	- MEMORY LEAK --> the third `ViewContainerRef` should be garbage collected, as there are no references to it being held in user code

Now uncomment line 35 in app.component.ts that manually clears the `_viewContainerRef` property on `ViewRef` detach:

```ts
// Manually clear the view ref's assigned previously assigned view container
// --> Allows the `viewContainerRef` to be garbage collected.
(this.viewRef as any)._viewContainerRef = null;
```

Do the same steps as above.
After clicking on "detach and destroy", there will only be **two** `ViewContainerRef` instances which is expected behavior.

## 🌍  Your Environment

**Angular Version:**
10.0.14

### PR DESCRIPTION
**Pull request by crisbeto**
When we attach a `ViewRef` to a `ViewContainerRef`, we save a reference to the container onto the `ViewRef` so that we can remove it when the ref is destroyed. The problem is that if the container's `hostView` is destroyed first, the `ViewRef` has no way of knowing that it should stop referencing the container.

These changes remove the leak by not saving a reference at all. Instead, when a `ViewRef` is destroyed, we clean it up through the `LContainer` directly. We don't need to worry about the case where the container is destroyed before the view, because containers automatically clean up all of their views upon destruction.

Fixes #38648.

### CODE REVIEW

File: packages/core/src/render3/view_ref.ts [hunk 1]
```diff
@@ -65,14 +68,21 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
...
     if (this._appRef) {
       this._appRef.detachView(this);
-    } else if (this._viewContainerRef) {
-      const index = this._viewContainerRef.indexOf(this);
-
-      if (index > -1) {
-        this._viewContainerRef.detach(index);
+    } else if (this._attachedToViewContainer) {
+      const parent = this._lView[PARENT];
+      if (isLContainer(parent)) {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/view_ref.ts:73]: Just a quick comment: may be we can add an assert to throw in dev mode in case parent is not an `LContainer`?

Comment (crisbeto) [reply packages/core/src/render3/view_ref.ts:73]: I was considering it, but I didn't know whether the `LView` could've been removed by something else or whether there would be a case where the parent is another `LView`.

File: packages/core/src/render3/view_ref.ts [hunk 2]
```diff
@@ -271,11 +281,11 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
     checkNoChangesInternal(this._lView[TVIEW], this._lView, this.context);
   }
 
-  attachToViewContainerRef(vcRef: viewEngine_ViewContainerRef) {
+  attachToViewContainerRef() {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/view_ref.ts:284]: Just a quick comment: the name of this method is a bit misleading now (feels like argument is missing) :)
May be we can rename it to something like `markAsAttachedToViewContainer`?

Comment (crisbeto) [reply packages/core/src/render3/view_ref.ts:284]: You're right for the Ivy case, but the method comes from an interface that is shared with VE. VE still uses the first argument so I think that the name is valid.

File: goldens/circular-deps/packages.json [hunk 3]
```diff
@@ -108,45 +108,6 @@
...
-  [
-    "packages/core/src/change_detection/change_detection.ts",
-    "packages/core/src/change_detection/change_detector_ref.ts",
-    "packages/core/src/render3/view_ref.ts",
-    "packages/core/src/linker/view_container_ref.ts",
-    "packages/core/src/render3/instructions/shared.ts",
-    "packages/core/src/error_handler.ts",
-    "packages/core/src/errors.ts",
-    "packages/core/src/view/types.ts",
-    "packages/core/src/linker/component_factory.ts"
```
Comment (mhevery) [inline goldens/circular-deps/packages.json:148]: Always a good sign when bunch of circular refs get removed!

❤️

**Review by mhevery — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: circular-dependencies

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:350578785:BASE:350578809:1610040775264:7abe1847).
